File: Array/Python/_0496_NextGreaterElementI.py
from typing import List


class Solution:
    # Iteration
    # Time Complexity: BigO(N^2)
    # Space Complexity: BigO(N)
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        dict = {}
        res = []

        for i in nums2:
            print('i', i)
            while len(stack) > 0 and i > stack[-1]:
                print('stack:', stack)
                print('add', i, 'to dict')
                dict[stack.pop()] = i
                print('dict:', dict)
            print('add', i, 'to stack')
            stack.append(i)

        for i in nums1:
            res.append(dict.get(i, -1))

        return res


ins = Solution()

print("Expected: [ -1, 3, -1 ]", "Output:",
      ins.nextGreaterElement([4, 1, 2], [1, 3, 4, 2]))
print("Expected: [ 3, -1 ]", "\tOutput:",
      ins.nextGreaterElement([2, 4], [1, 2, 3, 4]))
print("Expected: [7,7,7,7,7]", "\tOutput:",
      ins.nextGreaterElement([1, 3, 5, 2, 4], [6, 5, 4, 3, 2, 1, 7]))

File: Array/TypeScript/_0496_NextGreaterElementI.ts
/**
 * Iteration
 * Time Complexity: BigO(N^2)
 * Space Complexity: BigO(N)
 */
function nextGreaterElement(nums1: number[], nums2: number[]): number[] {
  let res: number[] = [];
  for (let i = 0; i < nums1.length; i++) {
    let idx = nums2.findIndex((num) => num == nums1[i]);
    let gt = nums2.filter((num, index) => num > nums1[i] && index > idx);
    if (gt.length > 0) res.push(gt[0]);
    else res.push(-1);
  }
  return res;
}

/**
 * Iteration
 * Time Complexity: BigO(N^2)
 * Space Complexity: BigO(N)
 */
function nextGreaterElementV2(nums1: number[], nums2: number[]): number[] {
  let res: number[] = [];
  nums1.forEach((num1) => {
    let idx = nums2.indexOf(num1);
    for (let j = idx + 1; j < nums2.length; j++) {
      if (nums2[j] > num1) {
        res.push(nums2[j]);
        return;
      }
    }
    res.push(-1);
  });
  return res;
}

console.log(
  "Expected: [ -1, 3, -1 ]",
  nextGreaterElementV2([4, 1, 2], [1, 3, 4, 2])
);
console.log("Expected: [ 3, -1 ]", nextGreaterElementV2([2, 4], [1, 2, 3, 4]));
console.log(
  "Expected: [7,7,7,7,7]",
  nextGreaterElementV2([1, 3, 5, 2, 4], [6, 5, 4, 3, 2, 1, 7])
);
